libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h

```diff
@@ -297,25 +297,74 @@ class LC_CSR_CSC_Labeled_Graph
     return (edgeLabelToIndexMap.find(data) != edgeLabelToIndexMap.end());
   }
 
+  enum FindEdgeWithLabelOptions {
+    FINDFIRST = 0,
+    FINDLAST,
+    FINDANY,
+  };
+
   /**
    * Returns an edge iterator to an edge with some node and key with some label
-   * by searching for the key via the node's outgoing or incoming edges.
+   * by searching for the key via the node's outgoing or incoming edges. By
+   * supplying find_first or find_last we can find the first or last satisfying
+   * edge. If neither are provided we will return an arbitrary satisfying edge.
+   *
+   * The return value is a tuple of <edge, lower bound for searching for last
+   * element, upper bound for searching for last element>
+   * if opt != FINDFIRST, the second two elements are arbitrary.
+   *
    * If not found, returns nothing.
    */
-  template <bool in_edges>
-  std::optional<edge_iterator> FindEdgeWithLabel(
-      GraphNode node, GraphNode key, const EdgeTy& data) const {
-    edge_iterator begin =
-        in_edges ? in_raw_begin(node, data) : raw_begin(node, data);
-    edge_iterator end = in_edges ? in_raw_end(node, data) : raw_end(node, data);
+  template <bool in_edges, enum FindEdgeWithLabelOptions opt>
+  std::optional<std::tuple<edge_iterator, edge_iterator, edge_iterator>>
+  FindFirstOrLastEdgeWithLabel(
+      const edge_iterator& begin, const edge_iterator& end, GraphNode node,
+      GraphNode key, const EdgeTy& data) const {
+    KATANA_LOG_DEBUG_ASSERT(!(find_first && find_last));
+
     edge_iterator l = begin;
     edge_iterator r = end - 1;
+
+    // If we're searching for the first edge, we stop at the beginning and need
+    // to check the previous edge. The opposite is true for the last edge.
+    edge_iterator limit = opt == FINDFIRST ? begin : (end - 1);
+    int direction = opt == FINDFIRST ? -1 : 1;
+
+    // If we're finding the first element, we can narrow the search space for
+    // finding the last element based off of what we've seen.
+    edge_iterator orig_end =
+        in_edges ? in_raw_end(node, data) : raw_end(node, data);
+    edge_iterator last_begin_bound = begin;
+    edge_iterator last_end_bound = end;
+
     while (r >= l) {
       edge_iterator mid = l + (r - l) / 2;
       GraphNode value =
           in_edges ? this->getInEdgeDst(mid) : this->getEdgeDst(mid);
       if (value == key) {
-        return mid;
+        // Update bounds for last only if we're searching for the first element.
+        if (opt == FINDFIRST) {
+          last_begin_bound = mid;
+          if (end != orig_end) {
```

Comment:
This is confusing. Shouldn't `last_end_bound = r` be correct? We know that the value of `r+1` is not the `key`.

I'd suggest dropping node from the parameters.

Reply:
yeah, that's true